runelite-client/src/main/java/net/runelite/client/plugins/fightcave/FightCaveOverlay.java

```diff
@@ -0,0 +1,181 @@
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package net.runelite.client.plugins.fightcave;
+
+import net.runelite.api.Client;
+import net.runelite.api.widgets.Widget;
+import net.runelite.api.widgets.WidgetInfo;
+import net.runelite.client.RuneLite;
+import net.runelite.client.ui.overlay.Overlay;
+import net.runelite.client.ui.overlay.OverlayPosition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.imageio.ImageIO;
+import java.awt.Color;
+import java.awt.Dimension;
+import java.awt.FontMetrics;
+import java.awt.Graphics2D;
+import java.awt.Image;
+import java.awt.Rectangle;
+import java.io.IOException;
+import java.io.InputStream;
+
+public class FightCaveOverlay extends Overlay
+{
+	private static final int WIDTH = 70;
+	private static final int SPACER = 6;
+	private static final int BOTTOM_BORDER = 4;
+	private static final Color GREEN_BACKGROUND = new Color(0, 255, 0, 100);
+	private static final Color RED_BACKGROUND = new Color(255, 0, 0, 100);
+
+	private static final Logger logger = LoggerFactory.getLogger(FightCaveOverlay.class);
+
+	private final FightCave plugin;
+	private final Client client = RuneLite.getClient();
+	private Image protectFromMagicImg;
+	private Image protectFromMissilesImg;
+
+	FightCaveOverlay(FightCave plugin)
+	{
+		super(OverlayPosition.DYNAMIC);
+		this.plugin = plugin;
+	}
+
+	@Override
+	public Dimension render(Graphics2D graphics)
+	{
+		JadAttack attack = plugin.getAttack();
+		Rectangle viewport = getViewportBounds();
+		if (attack == null || viewport == null)
+		{
+			return null;
+		}
+
+		Image img = getPrayerImage(attack);
+		if (img == null)
+		{
+			return null;
+		}
+
+		Color bgColor;
+		Color outlineColor;
+		Widget prayer = attack.getPrayerWidget(client);
+		FontMetrics fm = graphics.getFontMetrics();
+		int height = fm.getHeight() + img.getHeight(null) + SPACER + BOTTOM_BORDER;
+
+		if (client.isPrayerActive(attack.getPrayer()))
+		{
+			bgColor = GREEN_BACKGROUND;
+			outlineColor = Color.GREEN.darker();
+		}
+		else
+		{
+			bgColor = RED_BACKGROUND;
+			outlineColor = Color.RED.darker();
+		}
+
+		int bgX = (int) (viewport.getX() + viewport.getWidth() - WIDTH);
+		int bgY = (int) (viewport.getY() + viewport.getHeight() - height);
+		graphics.setColor(bgColor);
+		graphics.fillRect(bgX, bgY, WIDTH, height);
+
+		if (prayer != null)
+		{
+			graphics.setColor(outlineColor);
+			graphics.draw(prayer.getBounds());
+		}
+
+		graphics.setColor(Color.WHITE);
+		graphics.drawString("TzTok-Jad", bgX + (WIDTH - fm.stringWidth("TzTok-Jad")) / 2, bgY + fm.getHeight());
+		graphics.drawImage(img, bgX + (WIDTH - img.getWidth(null)) / 2, bgY + fm.getHeight() + SPACER, null);
+
+		return null;
+	}
+
+	private Rectangle getViewportBounds()
+	{
+		Widget viewport = getViewportWidget();
+		return viewport != null ? viewport.getBounds() : null;
+	}
+
+	private Widget getViewportWidget()
```

Comment:
This logic should probably be moved to somewhere more central. Maybe into the client mixin